### ISSUE TITLE
[restrict-template-expressions] option to ignore for console.log and similar methods

### ISSUE THREAD

**Issue by justingrant** [typescript-eslint/typescript-eslint#2261]
I just upgraded typescript-eslint and got many new `restrict-template-expressions` warnings, almost all of which were in `console.log` messages or other similar contexts where the warning was a false positive. In logging code, logging `undefined`/`null` as `"undefined"` or `"null"` is almost always the expected result, but in normal code the reverse is true. 

Therefore, I think this is a useful rule for "normal" code where strings are being constructed for showing to users, but if 99% of the time it's a false positive on log statements then it's hard to justify leaving it active in my projects.

Could typescript-eslint offer a way to automatically ignore this rule when the template strings are being constructed as parameters to `console.log`? Ideally, I could add a whitelist of other methods (e.g. custom loggers) whose parameters would also be ignored by this rule. 

I assume that the hard part would be uniquely identifying methods. IMHO I think it's OK to use dumb regex matching by the name used in the code, e.g. `methodName: [/\.log$/, /myCoolLogger/]` which puts the onus on developers to add any aliases.  A default could be `[/\.log$/, [/\.warn$/, [/\.info$/]` or maybe just `[/console\.$/]`.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/restrict-template-expressions": ["warn", { "allowNumber": true }]
  }
}
```

```TS
const foo = Math.random() < 0.5 ? 1 : undefined;
console.log(`foo is: ${foo}`);
```

**Expected Result**
There's some way for me to tell ESLint that some functions like `console.log` should allow `undefined` or `null` values-- if only because a big reason for logging is to report on unusual/unexpected values! 

**Actual Result**
```
Invalid type "1 | undefined" of template literal expression.
eslint@typescript-eslint/restrict-template-expressions
```

**Additional Info**
n/a

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.4.0` |
| `@typescript-eslint/parser`        | `3.4.0` |
| `TypeScript`                       | `3.9.5` |
| `ESLint`                           | `7.3.1` |
| `node`                             | `14.4.0` |
| `npm`                              | `6.14.5` |

**Comment by leethree** [comment]
This would be very helpful!

**Comment by OmgImAlexis** [comment]
Actually I think this comes down to more of how console should be used.
I don't actually see much of a reason to pass in a template sting when [format specifiers](https://github.com/nodejs/node/blob/master/doc/api/util.md#utilformatformat-args) exist.

Your example can be changed to this.
```ts
const foo = Math.random() < 0.5 ? 1 : undefined;
console.log('foo is: %s', foo);
```

**Comment by bradzacher** [comment]
I would agree.
Looking at this now - I don't think this is worthwhile when you can either use the format string version of `console.log` like you mentioned, or you can just add each as a separate argument and it'll log each one with a space.

the benefit of the latter is that each arg will be printed using the native formatter without worring about `%` commands.

![image](https://user-images.githubusercontent.com/7462525/106112375-07958380-6102-11eb-8ee4-7cc352c0d4b3.png)

**Comment by leethree** [comment]
template strings are easy to read and understand. it's not just used in `console.log` but also custom loggers. Besides, I don't think there is an industrial standard for "how console should be used."

> This deprives the ecosystem of the benefit of the rule being enabled on most code. This seems like a bad outcome for a good rule.

fully agreed.

**Comment by bradzacher** [comment]
The problem is that block lists are a bad way of configuring type-aware lint rules.

I've mentioned this countless times in issues in this repo.

A free text blocklist lacks any context, so it's easy to make it too wide.

For example, if you declare something like `methodName: ['log']`, then this will match any method called `log`.

Which means it'll match `Math.log`, which we all know is NOT a logging function. But the lack of context is what causes an incorrect match. Obviously `Math.log` is not a method that accepts a string, but this is just a simple example.

A blocklist that falsely matches creates false negatives.
False negatives are an insidious thing. They erode a user's trust in linting tools.
They're _**much**_ worse than false positives. False positives create noise that's annoying. Noise that's seen. Noise that can be grasped and understood. Noise that can be silenced.

False negatives are invisible. They lie in wait to cause a bug in production. Once discovered, the user asks "why didn't the linter catch this case?" followed by "what other cases have been missed?"
This makes the user second guess the linter, and trust it much less.

Sure - it's possible for us to build support in here specifically for `console` methods (it's a standard part of TS types, so a rule can interrogate the types and thus not false match).

But as mentioned - template strings will very easily cause you issues with many logging cases, so template strings are not a good way to do your logging. Which is why I would say that it's not a a false positive in the rule.

Would you rather have `[object Object]` in your production logs when you're investigating a bug all because you decided to ignore logging methods? 
Or would you rather have a properly logged object because the rule reported correctly and you ensured it was stringified?

**Comment by justingrant** [comment]
> The problem is that block lists are a bad way of configuring type-aware lint rules.
(snip)
> Sure - it's possible for us to build support in here specifically for `console` methods (it's a standard part of TS types, so a rule can interrogate the types and thus not false match).
(snip)
> Would you rather have `[object Object]` in your production logs when you're investigating a bug all because you decided to ignore logging methods?
> Or would you rather have a properly logged object because the rule reported correctly and you ensured it was stringified?

These concerns make sense. I'd be OK with override options that applied to `console` methods only, e.g. 

```
{
  allowNumber: true,
  allowBoolean: false,
  allowAny: false,
  allowNullish: false,
  console: {
    allowBoolean: true,
    allowNullish: true,
  }
}
```

It isn't a perfect solution (e.g. won't work for custom loggers) but it'd get rid of 95%+ of my false positives which in turn would let me justify leaving this rule enabled for all code.

**Comment by bradzacher** [comment]
happy to accept a PR for that sort of very limited scope option.

probably something like this in case we decide to add more overrides in later:
```js
{
  allowNumber: true,
  allowBoolean: false,
  allowAny: false,
  allowNullish: false,
  overrides: {
    console: {
      allowBoolean: true,
      allowNullish: true,
    }
  }
}
```

**Comment by monfera** [comment]
Linting rules toward structural typing become more and more elaborate, as they have a fractal nature. There's always more and more intricacies to consider. It's a misnomer if `restrict-template-expressions` has jurisdiction over `console.log` anyway. Also, it's perfectly OK for numbers etc. to appear in template strings. 

It feels there's a growing imbalance between ever more granular ways of constraining legal, standardized JavaScript by non-standards informal efforts. Is it good to add a _runtime_ `${String(myNumber)}` call into the template to placate some _static_ type checker? That's the default semantics. If this is deemed super useful, we ought to use a TS, compile-time annotation to indicate, "yes I know it is / can be a number and it's okay". 

Meanwhile, assuming no elaborate branding setup, TS and its linters don't bat an eyelid when the code adds an angle in radians to an angle in degrees 🤷‍♂️

**Comment by bradzacher** [comment]
The reason the rule can restrict numbers is because it's not always a good idea to dump a raw number into a string.

For example:
- would you rather dump `9.12345678912344578919372819` or would you rather dump `9.12`?
- would you rather dump `1627571085`, or would you rather dump `Thursday, 29 July 2021 3:04:45 PM`?
- would you rather dump `1000000000` or would you rather dump `1 billion` or maybe `1,000,000,000`?
- would you rather dump `3.14159265359` or would you rather dump `pi` or maybe `180 degrees`?

It's going to depend on your application as to exactly what representation you want a number. The idea is simply that often you will want to format a number before displaying it to a user.

The problem we have with static analysis is that it is missing contextual knowledge. It doesn't understand that in the context of a web application, a template string within a `console.log` statement is not actually going to be seen by a user, so it's okay to be more relaxed. The only way to make it understand that is to have someone teach it by hand.

**Comment by alfaproject** [comment]
Allowing string arrays would also be nice or other `toString` allowed types, maybe with a whitelist of the type name?

**Comment by bradzacher** [comment]
At this stage, I'm going to close this.
After over 2 years this issue has received a total of 5 reactions, and has not received additional feedback furthering - signalling that it's not something the majority of the community would like.
